File: PHP/bowling.php
<?php
define("PINS",10);
define("N_FRAMES",10);
define("STRIKE_SYMBOL" ,'X');
define("STRIKE_SYMBOL_LC" ,'x');
define("SPARE_SYMBOL",'/');
define("NO_POINTS_SYMBOL",'-');

// logical or operator is 'or' and not the double pipe || as we saw so far
function is_strike ($c) { return (($c == STRIKE_SYMBOL) or ($c==STRIKE_SYMBOL_LC)); }
function is_spare  ($c) { return ($c == SPARE_SYMBOL); }
function is_miss   ($c) { return ($c == NO_POINTS_SYMBOL); }

//we don't need this function, PHP has already is numeric
//function is_number ($c) { return is_numeric($c); }

//an interface is a trick used in java and 
class Frame implements ArrayAccess
{
    protected  $frameNumber;
    protected  $rolls;
    private    $score;

    //public is by default
    function __construct($frameNumber)
    {
        $this->frameNumber = $frameNumber;
        $this->rolls = array_fill(0,2,0);
        $this->score = 0;
    }

    public function getFrameNumber() {return $this->frameNumber;}

    //these are the methods needed to be implemented in the interface
    //this method is called when isset($obj[0]) is called
    public function offsetExists($idx){return isset($this->rolls[$idx]);}
    
    //this method is called in an expression like $obj[0]
    public function offsetGet($idx) {return $this->rolls[$idx];}

    //this method is called in an expression like $obj[0] = 1
    public function offsetSet($idx, $value) {$this->rolls[$idx] = $value; }

    //this method is called when unset($obj[0]) is invoked
    //In our case, we cannot lalow unset like that 
    //However, we can take advantage of this that, instead of removing the item
    //from the array, we set it to 0
    public function offsetUnset($idx) {$this->rolls[$idx] = 0;}  
    
    public function getScore() {return $this->score; }
    protected function setScore($score) {$this->score = $score;}
    
    public function isStrike() {return $this[0] == PINS;}
    public function isSpare() {return !$this->isStrike() && (($this[0]+$this[1])==PINS);}

    public function calculateScore($previous_frame, $next_rolls)
    {
        $frame_score = 0;

        if (isset($previous_frame)) $frame_score += $previous_frame->getScore();

        $frame_score += $this[0] + $this[1];

        if (count($next_rolls)>=2)
        {
            if ($this->isStrike()) $frame_score += $next_rolls[0] + $next_rolls[1]; 
            elseif ($this->isSpare()) $frame_score += $next_rolls[0]; 
        }

        $this->setScore($frame_score);
    }

    public function __toString()
    {
        //this little friend comes back :) - I love printf and its variations
        return sprintf("%s %s",$this->printRoll(0), $this->printRoll(1));
    }

    protected function printRoll($idx)
    {
        if (($idx != 1) and ($this[$idx] == PINS)) return STRIKE_SYMBOL;
        elseif (($idx==1) and ($this[$idx-1] ==PINS) ) return ' ';
        elseif (($idx>0) and ( ($this[$idx-1] + $this[$idx]) ==PINS) ) return SPARE_SYMBOL;
        elseif ($this[$idx] == 0) return NO_POINTS_SYMBOL;
        else return strval($this[$idx]);
    }
}

class LastFrame extends Frame
{
    //php here complains with this nonsense error:
    //PHP Fatal error:  Declaration of LastFrame::__constructor() must be compatible with Frame::__constructor($frameNumber)
    //I put a dummy parameter that it's never used
    //Finally, we are using parameters with default values
    //I need to do this because PHP is more pedantic that I remembered
    //PHP Fatal error:  Uncaught ArgumentCountError: Too few arguments to function LastFrame::__construct(), 0 passed
    public function __construct($_ = NULL)
    {
        //scope resolution operator ::
        //clearly borrowed from c++
        parent::__construct(N_FRAMES);
        $this->rolls = array_fill(0,3,0);
    }

    //same thing happens here - php not only does not overloading, it even dislike it!
    //Since I can call variables as I like, I simply call it _ (an underscore)
    public function calculateScore($previousFrame, $_)
    {
        $frame_score = $previousFrame->getScore();

        $frame_score += $this[0] + $this[1]; 

        if ($this->isStrike() or $this->isSpare()) $frame_score += $this[2];


        $this->setScore($frame_score);
    }

    public function __toString()
    {
        $buffer = "";
        if ($this->isSpare())
        {
            //this is the concatenation operator for string
            $buffer.=parent::__toString(); //here ToString must be called explicitily because base seems to be seen as a keyword rather than a regular object.
            $buffer.=" ";
            $buffer.=$this->printRoll(2);
        }
        elseif ($this->isStrike())
        {
            $buffer.=$this->printRoll(0);
            $buffer.=" ";
            $buffer.=$this->printRoll(1);
            $buffer.=" ";
            $buffer.=parent::printRoll(2);
        }
        else
        {
            $buffer.=parent::__toString(); 
            $buffer.="  ";
        }

        return $buffer;
    }

    protected function printRoll($idx)
    {
        if ($this[$idx] == 0) return NO_POINTS_SYMBOL;
        elseif ($this[$idx] == PINS) return STRIKE_SYMBOL;
        else return strval($this[$idx]);
    }
}

class Game implements ArrayAccess
{
    private $frames;
    public $player;
    
    public function __construct ($sequence,$player)
    {
        $this->frames = array();

        for($i = 0;$i<(N_FRAMES-1);$i++) $this->frames[$i] = new Frame($i+1);
        $this->frames[N_FRAMES-1] = new LastFrame();

        if (isset($sequence))
        {
            $this->readSequence($sequence);
            $this->calculateScore();
        }

        //wtf is this operator?!
        $this->player= $player??"Player";
    }

    //these are the methods needed to be implemented in the interface
    //this method is called when isset($obj[0]) is called
    public function offsetExists($idx){return isset($this->frames[$idx]);}
    
    //this method is called in an expression like $obj[0]
    public function offsetGet($idx) {return $this->frames[$idx];}

    //we cannot allow a set of a new frame externally. Here I prefer throwing an exception
    public function offsetSet($idx, $value) 
    {
        throw BadMethodCallException("This is a readonly property.");
    }

    //this method is called when unset($obj[0]) is invoked
    //In our case, we cannot lalow unset like that 
    //However, we can reset the frame using the offsetUnset method in Frame
    //I won't do this now because php has a limitation that python doesn't
    //The discussion will be moved then
    public function offsetUnset($idx) {throw BadMethodCallException("This is a readonly property.");}  

    public function readSequence($sequence)
    {
        $f = 0; 
        $r = 0;

        //php doesn't expose that a string is an array of characters.
        //therefore, to access each character in a string I need to split it to chunks of size 1
        $chars = str_split($sequence,1); 

        foreach($chars as $c)
        {
            if ($f < (N_FRAMES-1)) 
            {
                if (is_numeric($c) || is_miss($c)) 
                {
                    if (is_miss($c)) $this[$f][$r] = 0;
                    else $this[$f][$r] = intval($c);

                    $r++;
                    if ($r>1)
                    {
                        $f++;
                        $r=0;
                    }
                }
                else 
                {
                    if (($r==0) && is_strike($c)) 
                    {
                        $this[$f][0] = PINS;
                        $f++;
                    }
                    else
                    {
                        if (($r==1) && is_spare($c)) 
                        {
                            
                            $this[$f][1] = PINS - $this[$f][0];

                            $f++; 
                            $r=0;
                        }
                        else 
                        {
                            throw Exception(sprintf("I don't undernstanad character %c at position %d",$c,$i+1));
                        }
                    }
                }
            }
            else 
            {
                if (is_numeric($c) || is_miss($c)) 
                {
                    if (is_miss($c)) $this[$f][$r] = 0;
                    else $this[$f][$r] = intval($c);
                } 
                else if(is_strike($c)) $this[$f][$r] = PINS;
                else
                {

                    if (($r>0) && (is_spare($c))) $this[$f][$r] = PINS - $this[$f][$r-1];
                    else throw Exception(sprintf("I don't undernstanad character %c at position %d",$c,$i+1));
                    
                }

                $r++; 
            }
        }
    }

    public function calculateScore()
    {
        for ($i=0;$i<N_FRAMES;$i++)
        {
            $current_frame = $this[$i];
            $next_rolls = array_fill(0,2,0);

            if ($current_frame->getFrameNumber() < N_FRAMES) 
            {
                $next_rolls[0] = $this[$i+1][0];

                if ($this[$i+1]->getFrameNumber() < N_FRAMES)
                    $next_rolls[1] = $this[$i+1]->isStrike() ? $this[$i+2][0] : $this[$i+1][1];
                else $next_rolls[1] = $this[$i+1][1];                
            }
            
            
            if ($current_frame->getFrameNumber() > 1) $current_frame->calculateScore($this[$i-1],$next_rolls);
            else $current_frame->calculateScore(NULL,$next_rolls);
        }
    }

    public function printGame()
    {
        $frame_line1 = "";
        $frame_line2 = "";        

        foreach ($this->frames as $current_frame)
        {
            $frame_line1 .= strval($current_frame);
            $frame_line1 .= "|";

            switch($current_frame->getFrameNumber())
            {
                case N_FRAMES:
                    $frame_line2.= sprintf("%5d",$current_frame->getScore());
                    break;
                default:
                    $frame_line2.= sprintf("%3d",$current_frame->getScore());
            }

            $frame_line2 .= "|";

        }

        $dashed_line = str_repeat('-',strlen($frame_line1)+1)."\n";
        $player_line = $this->player;

        echo $dashed_line;
        
        echo "|". str_pad($this->player,strlen($frame_line1)-1," ", STR_PAD_BOTH) ."|\n";
               
        //print the top row of the scoring board - just the number of frames
        for($i=0;$i<N_FRAMES;$i++) echo "| ".strval($i+1)." ";
        echo " |\n";  
        
        echo $dashed_line;
        echo "|".$frame_line1."\n";

        echo $dashed_line;
        echo "|".$frame_line2."\n";
        
        echo $dashed_line;
    }
}

$my_game = "9-XX9-7/XX8-6/XX9";
$valerio = new Game($my_game,"Valerio");
$valerio->printGame();
?>

File: Python/bowling.py
allowed_roll = lambda lst,x : (x in lst) or (0<=int(x)<Roll.PINS)
#Yes, a new class could have been used, but I prefer this to introduce you to dictionaries
new_frame = lambda : {'printable_rolls':[],'rolls':[],'score':0,'cum_score':0}

is_strike = lambda r : isinstance(r,Strike)

class Roll:
    PINS = 10

    def __init__(this, value, look_ahead = 0):
        this.value = value
        this.look_ahead = look_ahead
    
    def __str__(this):
        return str(this.value) if (this.value>0) else '-'
    
    def __repr__(this):
        return str(this)
    
    def __add__(this,obj):
        if isinstance(obj,Roll) or (type(obj) == int):
            return this.value + int(obj)
    
    __radd__ = __add__
    
    def __int__(this):
        return this.value

class Spare(Roll):

    def __init__(this,previous_roll):
        super().__init__(Roll.PINS - previous_roll.value,1)
    
    def __str__(this):
        return '/'

class Strike(Roll):

    def __init__(this):
        super().__init__(10,2)
    
    def __str__(this):
        return 'X'

def convertToRoll(char, prev_roll=None):
    if (char in ['X','x']): return Strike()
    elif (char == '/'): return Spare(prev_roll)
    elif (char == '-'): return Roll(0)
    else: return Roll(int(char))


class Game:
    FRAMES = 10

    def __init__(this, sequence=None, player="Player"):
        this.__rolls = [] #this is python way to make private stuff
        this.player = player
        this.score = 0

        if (sequence is not None):
            this.readSequence(sequence)
            #this.calculateScore()

    #another magic method (__init__ is one)
    #this allows to use an instance as it was an array
    def __getitem__(this, key):
        return this.__rolls[key]

    def getFrames(this):
        prev_frame = None
        frame = new_frame()
        n_frame=1

        for i,roll in enumerate(this.__rolls):
            frame["printable_rolls"].append(str(roll))
            frame["rolls"].append(int(roll))
            frame["score"] += roll.value

            if ( (n_frame < Game.FRAMES) and ((is_strike(roll)) or (len(frame['rolls'])==2)) ):
                frame["score"] += sum(this[i+1:(i+roll.look_ahead+1)])
                frame['cum_score'] = frame['score'] + prev_frame['cum_score'] if (prev_frame is not None) else frame['score']

                yield frame

                prev_frame = frame
                frame = new_frame()
                n_frame+=1
            else: 
                if ((len(frame['rolls'])==2) and (frame["score"]<Roll.PINS) or #yields 10th frame if we have two rolls and cannot have more
                    (len(frame['rolls'])==3)): #yields 10th frame when we have 3 rolls
                    
                    #here we could avoid copy n paste somehow (ex nested function, but let's not overdo)
                    frame['cum_score'] = frame['score'] + prev_frame['cum_score'] if (prev_frame is not None) else frame['score']

                    yield frame


    def readSequence(this,sequence):
        this.__rolls = []

        frame = 1
        current_roll=0
        previous_roll = None

        for i,c in enumerate(sequence):
            roll = None
            if (current_roll == 0):
                allowed_symbols = ['-','X','x']
            else:
                if ((frame < Game.FRAMES) or ((frame == Game.FRAMES) and not is_strike(previous_roll)) ):
                    allowed_symbols = ['-','/']
                else:
                    allowed_symbols = ['-','X','x']
            
            if allowed_roll(allowed_symbols, c):
                roll = convertToRoll(c,previous_roll)
            else: raise Exception(f"I don't understand character {c} at position {i+1}")

            if (frame == Game.FRAMES):
                roll.look_ahead = 0 #strikes and spares in the last frame are just 10

            this.__rolls.append(roll)
            previous_roll = roll

            if (frame < Game.FRAMES):
                if (((current_roll==0) and (roll.value == 10)) or #move to next frame if we scored a strike not in the last frame
                    (current_roll>0)):                            #or when we are in the 2nd roll in any non-last frame
                    frame+=1
                    current_roll = 0
                else:
                    current_roll+=1

            else: #this is the last frame
                if (current_roll<2):
                    current_roll+=1

                    if (sum(this[-2:])<Roll.PINS): #we are not entitled to the extra roll(s)
                        break #exit for prematurely
                else:
                    break #we reached the last roll
    
    def __str__(this):
        frame_line0 = []
        frame_line1 = []
        frame_line2 = []

        for i,frame in enumerate(this.getFrames()):
            size = 3 if i<(Game.FRAMES-1) else 5
            
            frame_line0.append(str(i+1).center(size))
            frame_line1.append(" ".join(frame['printable_rolls']).ljust(size))
            frame_line2.append(str(frame['cum_score']).rjust(size))
        
        line0= "|{}|".format("|".join(frame_line0))
        line1= "|{}|".format("|".join(frame_line1))
        line2= f"|{'|'.join(frame_line2)}|" #let's use a different string formatting approach

        dashed_line = "-"*len(line0)

        formatted_player = "|"+ this.player.center(len(line0)-2) +"|"

        return f"{dashed_line}\n{formatted_player}\n{line0}\n{dashed_line}\n{line1}\n{dashed_line}\n{line2}\n{dashed_line}\n"

    __repr__ = __str__

my_game = "9-XX9-7/XX8-6/XX9"
valerio = Game(my_game,"Valerio")
print(valerio)